### PR TITLE
Correct parsing string interpolation with trailing newline

### LINKED ISSUE #4713: Unterminated string interpolation (parser bug)
Admittedly this is a contrived example, but it's based on an actual thing i was doing. just trying to get you the simplest working demonstration of the bug:

This works:
```
arry = ["a", "b", "c"]

puts "my array #{arry.map{|c|c.to_s}}"
```

This does not:
```
arry = ["a", "b", "c"]

puts "my array #{arry.map{|c|c.to_s}
}"
```

This is unrelated to any whitespace (or lack thereof) at the start of the last line.
```
$ crystal -v
Crystal 0.22.0 (2017-04-20) LLVM 4.0.0
```

### PR DESCRIPTION
**Pull request by MakeNowJust**
Fixed #4713

### CODE REVIEW

**Comment by RX14** [comment]
This fixes parsing 
```cr
puts "foo #{"bar"
}"
```

to produce `foo bar` in stdout, however that example still can't be formatted.

If we can come to a *quick* consensus on how this should be formatted we should include that change in this PR, if it gets to be a lengthy discussion we can fix it later.

**Comment by RX14** [comment]
Also "Correctly parse ..." is correct English, instead of "Correct to parse ..."

**Comment by straight-shoota** [comment]
I am not sure if the formatter should remove newlines even in the first case.

**Comment by MakeNowJust** [comment]
@straight-shoota Even if the first case's newline is intended, we can use backslash + newline escape in this case (and I think we should use it.)

```crystal
puts "foo #{"bar"}\
"
```

**Comment by MakeNowJust** [comment]
@RX14 I can't understand the difference between "to do" and "doing"...

**Comment by MakeNowJust** [comment]
I considered some ideas to format string interpolation with newline:

Original source code:

```crystal
"foo #{
  bar = 1
} baz"
```

Idea 1: keep

```crystal
"foo #{
  bar = 1
} baz"
```

Idea 2: align with delimiter

```crystal
"foo #{
       bar = 1
     } baz"
```

Idea 3: remove newline and indent

```crystal
"foo #{bar = 1} baz"
```

I don't know which is the best and I can't decide it. I guess, the preference of idea 1 is unstable (and I don't like it). Idea 2 is the best if it follows no string, so:

```crystal
"foo #{
       bar = 1
     }"
```

this indentation looks like `foo = begin ... end` pattern, so it is good enough. However when it follows any string, I think it is ugly. Idea 3, `crystal tool format` disallow newline inside string interpolation in other words, is good for me, because I think we shouldn't write too complex expression to need newline inside string interpolation, but I don't know this is general opinion.

**Comment by RX14** [comment]
I think the rule should be that if there's a newline after `#{` we do idea 1 (indent 1 indent then format normally), otherwise keep it as-is. That means that the original example formats to

```cr
puts "foo #{"bar"}"
```

but 

```cr
puts "foo #{
  "bar"
}"
```

formats as-is.

**Comment by MakeNowJust** [comment]
Ok. I implemented formatting string interpolation in this way.

File: spec/compiler/formatter/formatter_spec.cr [hunk 1]
```diff
@@ -340,6 +340,11 @@ describe Crystal::Formatter do
   assert_format "__DIR__", "__DIR__"
   assert_format "__LINE__", "__LINE__"
 
+  assert_format "\"\#{foo = 1\n}\"", "\"\#{foo = 1}\""
```
Comment (bew) [inline spec/compiler/formatter/formatter_spec.cr]: You could use `%(...)` to make the assertion more readable ?

e.g:
```cr
assert_format %("\#{foo = 1\n}"), %("\#{foo = 1}")
```

Comment (MakeNowJust) [inline spec/compiler/formatter/formatter_spec.cr]: fixed

**Comment by mverzilli** [comment]
Thank you @MakeNowJust!

**Pull request merged**